### PR TITLE
Even better docs

### PR DESCRIPTION
**Pull request by glwagner**
This PR rewrites much of the docs, adds a show function for `Model` and various boundary conditions structures (which fixes very long output in the docs), and updates many of the examples to be docs-friendly.

The docs are still very much preliminary even after the updates in this PR. Once this PR is merged, I will open an issue to define a roadmap for getting the docs to the place we need them to be.

### CODE REVIEW

File: src/models.jl [hunk 1]
```diff
@@ -84,6 +84,37 @@ function Model(;
...
+    else
+        return string(typeof(dict), " with $(length(dict)) entries:", '\n',
+                      Tuple(string(padchar, 
+                                   "   ├── ", name, " => ", typeof(dict[name]), '\n') 
+                            for name in dict.keys[1:end-1]
+                           )...,
+                           padchar, "   └── ", dict.keys[end], " => ", typeof(dict.vals[end])
+                      )
+    end
+end
```
Comment (ali-ramadhan) [inline src/models.jl]: Probably belongs in `utils.jl`?

**Comment by ali-ramadhan** [comment]
Examples starting to look great!

We should probably add `show` functions for a few more structs like `JLD2OutputWriter` but maybe there's a way to suppress the printing? I tried adding a semicolon last weekend but that didn't help. I'll ask on the `#documentation` channel.

Few more comments on the examples:
1. In `ocean_convection_with_plankton.jl` you say "Finally, we end the simulation after 1 day." but `end_time = day / 2`. Might also be good to explain that Oceananigans exports `day`. I should probably add docstrings to `second`, `minute`, `hour`, `day`.

2. Might be better to use `pcolormesh` or `contourf` for the `two_dimensional_turbulence.jl` example plotting.

3. Also might be good to add colorbars to the plots in the examples, but this can wait.

4. All the colormaps are `viridis`. Again, this can wait.

**Comment by ali-ramadhan** [comment]
We should also figure out how to get equation numbers and `\eqref` to work with KaTeX. KaTeX has pretty good documentation so shouldn't be too hard.

But if KaTeX is limiting us, we can switch back to the MathJax backend through Documenter.

image

File: docs/src/manual/physics.md [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+
+The Coriolis model controls the manifestation of the term $\bm{f} \times \bm{u}$ in the momentum equation.
+
+### The "$f$-plane" approximation
+
+Under an $f$-plane approximation[^3] the reference frame in which 
+the momentum and tracer equations are are solved rotates at a constant rate around a 
+vertical axis, such that 
+```math
+    \bm{f} \equiv f \bm{\hat z}
```
Comment (ali-ramadhan) [inline docs/src/manual/physics.md]: Why use `=` for the beta-plane definition but `\equiv` for the f-plane?

File: docs/make.jl [hunk 3]
```diff
@@ -40,38 +38,38 @@ makedocs(
...
                  "Taylor-Green vortex"     => "verification/taylor_green_vortex.md",
-                 "Lid-driven cavity"       => "verification/lid_driven_cavity.md",
-                 "Free convection"         => "verification/free_convection.md",
                  "Stratified Couette flow" => "verification/stratified_couette_flow.md"
-             ]
+             ],
+             #"Appendix" => [
+                 #"Staggered grid"         => "manual/staggered_grid.md",
+                 #"Fractional step method" => "manual/fractional_step.md",
+             #],
```
[the remark this answers is not in the document]
Comment (glwagner) [reply docs/make.jl]: I don’t think they are helpful to users yet. They need a bit of work.

**Comment by glwagner** [comment]
KaTeX cannot do equation numbers or references yet. There are a number of issues open on katex github about this. 

I think to suppress output you’d have to add a line with ‘nothing’. But this is silly, so yes, show for the output writers is important. 

I tried to keep the plotting simple and barebones so 1. we don’t distract from the point of the example, which is to demonstrate Oceananigans usage (not to demonstrate plotting) and 2. so the examples are easy to maintain. I agree that colorbars are useful for interpreting output, but we don’t interpret output in the examples. 

I think viridis is an acceptable colormap, but I’m also happy to pick a prettier one. As I noted above, I think we should keep plotting simple and involving as few characters as possible so the examples are maximally useful in demonstrating how Oceananigans is used. (There’s really too much code devoted to plotting as is, which we should fix by shipping plotting recipes with Oceananigans.)

The package name is Oceananigans.jl. 

Typos noted, I will fix before merging.

File: docs/src/manual/physics.md (hunk 2 above)
Comment (glwagner) [reply docs/src/manual/physics.md]: Its a typo

File: docs/make.jl (hunk 3 above)
[the remark this answers is not in the document]
Comment (ali-ramadhan) [reply docs/make.jl]: That's more a matter of opinion.

But my name is not the only one appearing on the docs so I'll work on them until everyone is happy with the appendices.

**Comment by ali-ramadhan** [comment]
> I tried to keep the plotting simple and barebones so 1. we don’t distract from the point of the example, which is to demonstrate Oceananigans usage (not to demonstrate plotting) and 2. so the examples are easy to maintain. I agree that colorbars are useful for interpreting output, but we don’t interpret output in the examples.

True but using `pcolormesh` or `contourf` with a colorbar adds at most a few lines. I believe most people don't actually read examples from start to finish and just quickly scroll through the text to skim the code and any plots. So having a plot with some labels might help to these potential users won't have to search the text to figure out what they're looking at.

> There’s really too much code devoted to plotting as is, which we should fix by shipping plotting recipes with Oceananigans.

That might help for the complicated plots although I think it'll be useful to keep some introductory examples that use the regular plotting functions so users can see how to go about plotting their own simulations without the use of our plotting recipes.

> I thought the package name is Oceananigans.jl? What is convention?

Maybe getting rid of the backticks might help as it looks pretty distracting in the docs with backticks:
image

File: docs/make.jl (hunk 3 above)
[the remark this answers is not in the document]
Comment (glwagner) [reply docs/make.jl]: You're right in general that whether or not docs are more useful than harmful is an opinion... ! But I'm not talking about the content. I'm talking about the way that I chopped up the overleaf and pasted the text in. It's chaotic and random and there's no reference to it in the main part of the docs. It's just noise at this point, the basic work of shaping the content into an appendix and referencing it remains.

**Comment by glwagner** [comment]
> True but using pcolormesh or contourf with a colorbar adds at most a few lines (for the 2D turbulence example)

It was a conscious choice on my part to use `imshow` rather than `pcolormesh` (not laziness), but I suppose you disagree with my choice.

If "vorticity" as a title is not clear enough we can say "Heatmap of vorticity on an arbitrary scale". Would that be more clear?

**Comment by ali-ramadhan** [comment]
What you've done with the examples makes more sense now if viewed as utilitarian examples meant to show case how you can interact with the code.

For users who are trying to familiarize themselves with Julia, maybe it could help if we further split the time stepping and plotting so that all time stepping happens in the first `n` lines and afterwards some plotting happens so it's clearer to the user what is Oceananigans code and what is plotting code.

But yeah I'm happy with the examples in this PR. There will be plenty of opportunities to improve.

Should we merge this now? The docs have been improved, and would be good to start getting some feedback.

**Pull request merged**